File: Self-practice/treehugging.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    vector<set<pair<int, int>>> adj_list(2 * n);
    stack<int> leaves;

    for (int e = 0; e < 2 * (n - 1); e++) {
        int u, v;
        cin >> u >> v;
        
        adj_list[v].emplace(n + u, e);
        adj_list[n + u].emplace(v, e);
    }

    for (int u = 2; u < 2 * n; u++) {
        if (adj_list[u].empty()) {
            cout << "impossible";
            exit(0);
        }
        if (adj_list[u].size() == 1) leaves.emplace(u);
    }

    string s(2 * (n - 1), ' ');

    while (!leaves.empty()) {
        int u = leaves.top();
        leaves.pop();
        
        if (adj_list[u].empty()) {
            cout << "impossible";
            exit(0);
        }

        auto [v, e] = *adj_list[u].begin();
        adj_list[u].erase(adj_list[u].begin());
        adj_list[v].erase({u, e});
        s[e] = u <= n ? 'L' : 'R';
        if (adj_list[v].size() == 1) leaves.emplace(v);
    }

    for (int u = 2; u < 2 * n; u++)
        while (!adj_list[u].empty()) {
            auto [v, e] = *adj_list[u].begin();
            adj_list[u].erase(adj_list[u].begin());
            adj_list[v].erase({u, e});
            s[e] = u <= n ? 'L' : 'R';
            u = v;
        }

    cout << s;
}

File: Self-practice/treehugging.py
from io import StringIO

n = int(input())
adj_list = [set() for _ in range(2 * n)]
leaves = []

for e in range(2 * (n - 1)):
    u, v = map(int, input().split())
    adj_list[v].add((n + u, e))
    adj_list[n + u].add((v, e))

for u in range(2, 2 * n):
    if not adj_list[u]:
        print('impossible')
        exit(0)
    if len(adj_list[u]) == 1:
        leaves.append(u)

solution = StringIO()

while leaves:
    u = leaves.pop()
    if not adj_list[u]:
        print('impossible')
        exit(0)
    (v, e) = adj_list[u].pop()
    solution.seek(e)
    solution.write('L') if u <= n else solution.write('R')
    adj_list[v].remove((u, e))
    if len(adj_list[v]) == 1:
        leaves.append(v)

for u in range(2, 2 * n):
    while adj_list[u]:
        (v, e) = adj_list[u].pop()
        solution.seek(e)
        solution.write('L') if u <= n else solution.write('R')
        adj_list[v].remove((u, e))
        u = v

print(solution.getvalue())
